### PR TITLE
Thread safe fuse mounts.

### PR DESCRIPTION
**Pull request by ppknap**
This PR replaces FUSE's old thread-unsafe Node implementation with the safer one. 

## Motivation and Context
Mount could crash on high file load.

## How Has This Been Tested?
Manually

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: go/src/koding/klient/machine/index/node/node.go [hunk 1]
```diff
@@ -2,292 +2,358 @@ package node
...
+	return n != nil && n.Entry.Virtual.Promise.Exist()
+}
 
-func (ns NodeSlice) Len() int           { return len(ns) }
-func (ns NodeSlice) Less(i, j int) bool { return ns[i].Name < ns[j].Name }
-func (ns NodeSlice) Swap(i, j int)      { ns[i], ns[j] = ns[j], ns[i] }
+// Clone returns a deep copy of called node.
+func (n *Node) Clone() *Node {
+	c := &Node{
+		Name: n.Name,
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/index/node/node.go]: children: make([]*Node, 0, len(n.children)),

File: go/src/koding/klient/machine/index/node/tree.go [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+
+func defaultInodeIDGenerator() func() uint64 {
+	var current uint64 = RootInodeID
+	return func() uint64 {
+		inodeID := atomic.AddUint64(&current, 1)
+		if inodeID <= RootInodeID {
+			return atomic.AddUint64(&current, 1)
+		}
+
+		return inodeID
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/index/node/tree.go]: if inodeID at L19 is max uint64, then here it's RootInodeID

File: go/src/koding/klient/machine/index/node/tree.go [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+		subFn("", n)
+		return true
+	}
+}
+
+// ExistCount stores the number of nodes that are proven to exist.
+func ExistCount(n *int) Predicate {
+	return Walk(func(nd *Node) {
+		if nd.Entry != nil && nd.Entry.Virtual.Promise.Exist() {
+			(*n)++
```
Comment (rjeczalik) [inline go/src/koding/klient/machine/index/node/tree.go]: no need for parenthesis

**Review by rjeczalik — APPROVED** [review]

File: go/src/koding/klient/machine/index/node/tree.go (hunk 3 above)
Comment (ppknap) [inline go/src/koding/klient/machine/index/node/tree.go]: C/C++ habits :)

**Pull request merged**